derive-eip712/src/lib.rs

```diff
@@ -0,0 +1,399 @@
+//! # EIP-712 Derive Macro
+//! This crate provides a derive macro `Eip712` that is used to encode a rust struct
+//! into a payload hash, according to [https://eips.ethereum.org/EIPS/eip-712](https://eips.ethereum.org/EIPS/eip-712)
+//!
+//! The trait used to derive the macro is found in `ethers_core::transaction::eip712::Eip712`
+//! Both the derive macro and the trait must be in context when using
+//!
+//! This derive macro requires the `#[eip712]` attributes to be included
+//! for specifying the domain separator used in encoding the hash.
+//!
+//! All String values returned by the implemented methods are hex encoded and should be
+//! decoded into `[u8; 32]` for signing. See example for decoding.
+//!
+//! # Example Usage
+//!
+//! ```rust
+//! use derive_eip712::*;
+//! use ethers_core::types::{transaction::eip712::Eip712, H160};
+//! use serde::Serialize;
+//!
+//! #[derive(Debug, Eip712, Serialize)]
+//! #[eip712(
+//!     name = "Radicle",
+//!     version = "1",
+//!     chain_id = 1,
+//!     verifying_contract = "0x0000000000000000000000000000000000000000"
+//! )]
+//! pub struct Puzzle {
+//!     pub organization: H160,
+//!     pub contributor: H160,
+//!     pub commit: String,
+//!     pub project: String,
+//! }
+//!
+//! let puzzle = Puzzle {
+//!     organization: "0000000000000000000000000000000000000000"
+//!         .parse::<H160>()
+//!         .expect("failed to parse address"),
+//!     contributor: "0000000000000000000000000000000000000000"
+//!         .parse::<H160>()
+//!         .expect("failed to parse address"),
+//!     commit: "5693b7019eb3e4487a81273c6f5e1832d77acb53".to_string(),
+//!     project: "radicle-reward".to_string(),
+//! };
+//!
+//! let hash = puzzle.encode_eip712()?;
+//!
+//! let decoded: Vec<u8> = hex::decode(hash).expect("failed to decode")
+//! let byte_array: [u8; 32] = <[u8; 32]>::try_from(&decoded[..])?;
+//! ```
+//!
+//! # Limitations
+//!
+//! At the moment, the derive macro does not recursively encode nested Eip712 structs.
+//!
+
+use std::collections::HashMap;
+
+use convert_case::{Case, Casing};
+use proc_macro::TokenStream;
+use quote::quote;
+
+use ethers_core::{
+    abi,
+    abi::Token,
+    types::{Address, H160, U256},
+    utils::keccak256,
+};
+
+/// This method provides the hex encoded domain type hash for EIP712Domain type;
+/// This is used by all Eip712 structs.
+fn eip712_domain_type_hash() -> [u8; 32] {
+    keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)")
+}
+
+/// Eip712 Domain attributes used in determining the domain separator;
+#[derive(Debug, Default)]
+struct Eip712Domain {
```

Comment:
Move types to ethers_core::transaction::eip712 so they are available outside of the macro crate.

Reply:
moved.